csharp/ql/lib/semmle/code/csharp/commons/QualifiedName.qll

```diff
@@ -1,6 +1,201 @@
+
 /**
- * Provides predicates related to C# qualified names.
+ * Holds if namespace `n` has the qualified name `qualifier`.`name`.
+ *
+ * For example if the qualified name is `System.Collections.Generic`, then
+ * `qualifier`=`System.Collections` and `name`=`Generic`.
  */
+predicate namespaceHasQualifiedName(DotNet::Namespace n, string qualifier, string name) {
+  if n instanceof DotNet::GlobalNamespace
+  then qualifier = "" and name = ""
+  else (
+    exists(string pqualifier, string pname |
+      namespaceHasQualifiedName(n.getParentNamespace(), pqualifier, pname) and
+      qualifier = getQualifiedName(pqualifier, pname)
+    ) and
+    name = n.getName()
+  )
+}
+
+/** Gets a string of `N` commas where `N + 1` is the number of type parameters of this unbound generic. */
+private string getTypeParameterCommas(UnboundGeneric ug) {
+  result = strictconcat(int i | exists(ug.getTypeParameter(i)) | "", ",")
+}
+
+/** Provides the input to `QualifiedName`. */
+signature module QualifiedNameInputSig {
+  /** Gets the suffix to print after unbound generic `ug`. */
+  default string getUnboundGenericSuffix(UnboundGeneric ug) {
+    result = "<" + getTypeParameterCommas(ug) + ">"
+  }
+}
+
+/** Provides predicates for computing fully qualified names. */
+module QualifiedName<QualifiedNameInputSig Input> {
+  private string getDimensionString(ArrayType at, Type elementType) {
+    exists(Type et, string res |
+      et = at.getElementType() and
+      res = at.getArraySuffix() and
+      if et instanceof ArrayType
+      then result = res + getDimensionString(et, elementType)
+      else (
+        result = res and elementType = et
+      )
+    )
+  }
+
+  bindingset[t]
+  private string getFullName(Type t) {
+    exists(string qualifier, string name |
+      hasQualifiedName(t, qualifier, name) and
+      result = getQualifiedName(qualifier, name)
+    )
+  }
+
+  /** Gets the concatenation of the `getFullName` of type arguments. */
+  language[monotonicAggregates]
+  private string getTypeArgumentsQualifiedNames(ConstructedGeneric cg) {
+    result =
+      strictconcat(Type t, int i | t = cg.getTypeArgument(i) | getFullName(t), "," order by i)
+  }
+
+  /** Holds if declaration `d` has the qualified name `qualifier`.`name`. */
+  predicate hasQualifiedName(Declaration d, string qualifier, string name) {
+    d =
+      any(ValueOrRefType vort |
+        vort instanceof VoidType and
+        qualifier = "System" and
+        name = "Void"
+        or
+        vort =
+          any(ArrayType at |
+            exists(Type elementType, string name0 |
+              hasQualifiedName(elementType, qualifier, name0) and
+              name = name0 + getDimensionString(at, elementType)
+            )
+          )
+        or
+        hasQualifiedName(vort.(TupleType).getUnderlyingType(), qualifier, name)
+        or
+        qualifier = "System" and
+        name = "Nullable<" + getFullName(vort.(NullableType).getUnderlyingType()) + ">"
```

Comment:
Is this needed? `NullableType` is a subclass of `ConstructedType`, couldn't the below `any(ConstructedType ct | ...` cover nullable types too?

Reply:
Again, I just copied it from the existing setup, but I can try to remove it.